node/forest_libp2p/src/chain_exchange/behaviour.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use ahash::HashMap;
+use libp2p::{
+    request_response::{
+        OutboundFailure, ProtocolSupport, RequestId, RequestResponse, ResponseChannel,
+    },
+    swarm::NetworkBehaviour,
+    PeerId,
+};
+use log::warn;
+
+use super::*;
+use crate::{rpc::RequestResponseError, service::metrics};
+
+type InnerBehaviour = RequestResponse<ChainExchangeCodec>;
+
+pub struct ChainExchangeBehaviour {
+    inner: InnerBehaviour,
+    response_channels:
+        HashMap<RequestId, flume::Sender<Result<ChainExchangeResponse, RequestResponseError>>>,
+}
+
+impl ChainExchangeBehaviour {
+    pub fn send_request(
+        &mut self,
+        peer: &PeerId,
+        request: ChainExchangeRequest,
+        response_channel: flume::Sender<Result<ChainExchangeResponse, RequestResponseError>>,
+    ) -> RequestId {
+        let request_id = self.inner.send_request(peer, request);
+        self.response_channels.insert(request_id, response_channel);
+        self.track_metrics();
+        request_id
+    }
+
+    pub fn complete_request(&mut self, request_id: RequestId) {
+        self.response_channels.remove(&request_id);
+    }
+
+    pub fn send_response(
+        &mut self,
+        channel: ResponseChannel<ChainExchangeResponse>,
+        response: ChainExchangeResponse,
+    ) -> Result<(), ChainExchangeResponse> {
+        self.inner.send_response(channel, response)
+    }
+
+    pub async fn handle_inbound_response(
+        &mut self,
+        request_id: &RequestId,
+        response: ChainExchangeResponse,
+    ) {
+        if let Some(channel) = self.response_channels.remove(request_id) {
+            self.track_metrics();
+            if let Err(err) = channel.send_async(Ok(response)).await {
+                warn!("{err}");
+            }
+        }
+    }
+
+    pub fn on_outbound_error(&mut self, request_id: &RequestId, error: OutboundFailure) {
+        if let Some(tx) = self.response_channels.remove(request_id) {
+            if let Err(err) = tx.send(Err(error.into())) {
+                warn!("{err}");
+            }
+            self.track_metrics();
```

Comment:
Any particular reason why we track metrics here after the send and in `handle_inbound_response` before? If not, it'd be nice to keep the same flow.

Reply:
Fixed. Not sure why I did that : )